### PR TITLE
Introducing FlytePropeller TURBO Mode #minor

### PR DESCRIPTION
**Pull request by kumare3**
# TL;DR
This PR adds a new configuration to FlytePropeller which is enabled by default. This improves the transition latency of FlytePropeller by an order of magnitude or more. Absolute comparison to come soon.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Currently FlytePropeller event loop is designed to make the smallest forward progress in each round. There are cases in which the round progress is purely book-keeping. For example moving a workflow from accepted to Running state. Moving a node from new to queued to running state. This is done so that state is resilient from crashes.

the turbo mode, essentially waits for the one of these rounds to complete and optimistically progresses to the next round using the updated state. To add safety to the optimistic progress we can configure bounds.

The psuedo algorithm
```
w -> Original workflow state
w1 -> updated workflow state

while w1 != w and round < configured:
     w1 = try_mutate(w)
     if w1 != w:
         record(w1)
         w = w1
         round++
```

## Tracking Issue
https://github.com/flyteorg/flyte/issues/676

### CODE REVIEW

File: pkg/controller/config/config.go [hunk 1]
```diff
@@ -74,6 +74,7 @@ type Config struct {
 	MaxDatasetSizeBytes    int64                `json:"max-output-size-bytes" pflag:",Maximum size of outputs per task"`
 	KubeConfig             KubeClientConfig     `json:"kube-client-config" pflag:",Configuration to control the Kubernetes client"`
 	NodeConfig             NodeConfig           `json:"node-config,omitempty" pflag:",config for a workflow node"`
+	EnableFastFollow       bool					`json:"enable-fast-follow" pflag:",Boolean flag that enables Fast Follow mode, this makes Propeller proceed to another round on successful write to etcD."`
```
Comment (EngHabu) [inline pkg/controller/config/config.go:77]: nit: formatting seems to have more spaces... tabs?

File: pkg/controller/handler.go [hunk 2]
```diff
@@ -181,39 +181,53 @@ func (p *Propeller) Handle(ctx context.Context, namespace, name string) error {
...
-			// If the end result is a terminated workflow, we remove the labels
-			// We add a completed label so that we can avoid polling for this workflow
-			SetCompletedLabel(mutatedWf, time.Now())
-			ResetFinalizers(mutatedWf)
+		// TODO we will need to call updatestatus when it is supported. But to preserve metadata like (label/finalizer) we will need to use update
+
+		// update the GetExecutionStatus block of the FlyteWorkflow resource. UpdateStatus will not
+		// allow changes to the Spec of the resource, which is ideal for ensuring
+		// nothing other than resource status has been updated.
+		newWf, updateErr := p.wfStore.Update(ctx, mutatedWf, workflowstore.PriorityClassCritical)
```
Comment (EngHabu) [inline pkg/controller/handler.go:213]: Time to switch to UpdateStatus?

**Review by EngHabu — DISMISSED** [review]
Sweet!

File: pkg/controller/handler.go (hunk 2 above)
Comment (kumare3) [reply pkg/controller/handler.go:217]: I want to do that as a follow up. It is a little dangerous to do that at the moment, because we add a finalizer in some places. Do not want to de-stabilize

File: pkg/controller/config/config.go (hunk 1 above)
Comment (kumare3) [reply pkg/controller/config/config.go:79]: will fix formatting

File: pkg/controller/handler.go [hunk 3]
```diff
@@ -180,40 +186,62 @@ func (p *Propeller) Handle(ctx context.Context, namespace, name string) error {
...
-			if mutatedWf.Status.Equals(&w.Status) {
-				logger.Info(ctx, "WF hasn't been updated in this round.")
-				return nil
+	maxLength := p.cfg.MaxStreakLength
+	if maxLength <= 0 {
+		maxLength = 1
+	}
+
+	for streak = 0; streak < maxLength; streak++ {
+		mutatedWf, err := p.TryMutateWorkflow(ctx, w)
```
Comment (EngHabu) [inline pkg/controller/handler.go:198]: Do we measure round time inside TryMuttateWorkflow? won't this be a bit misleading? should we measure it here and record it `n` times instead?

File: pkg/controller/config/config.go [hunk 4]
```diff
@@ -74,6 +76,8 @@ type Config struct {
 	MaxDatasetSizeBytes    int64                `json:"max-output-size-bytes" pflag:",Maximum size of outputs per task"`
 	KubeConfig             KubeClientConfig     `json:"kube-client-config" pflag:",Configuration to control the Kubernetes client"`
 	NodeConfig             NodeConfig           `json:"node-config,omitempty" pflag:",config for a workflow node"`
+	EnableTurboMode        bool                 `json:"enable-turbo-mode" pflag:",Boolean flag that enables Turbo-mode, this makes Propeller proceed to another round on successful write to etcD."`
```
Comment (EngHabu) [inline pkg/controller/config/config.go:79]: Should we really have two flags? or just have it enabled all the time and use `MaxStreakLength <= 1` meaning it essentially has no effect?

File: pkg/controller/workflowstore/inmemory.go [hunk 5]
```diff
@@ -50,8 +50,10 @@ func (i *InmemoryWorkflowStore) UpdateStatus(ctx context.Context, w *v1alpha1.Fl
 		if w.Name != "" && w.Namespace != "" {
 			if m, ok := i.store[w.Namespace]; ok {
 				if _, ok := m[w.Name]; ok {
-					m[w.Name] = w
-					return w, nil
+					newW := w.DeepCopy()
+					newW.ResourceVersion = w.ResourceVersion + "x"
```
Comment (EngHabu) [inline pkg/controller/workflowstore/inmemory.go:54]: What's this?

Comment (kumare3) [reply pkg/controller/workflowstore/inmemory.go:54]: every successful write changes the resource version for local testing

File: pkg/controller/config/config.go (hunk 4 above)
Comment (kumare3) [reply pkg/controller/config/config.go:79]: that is interesting

File: pkg/controller/handler.go (hunk 3 above)
Comment (kumare3) [reply pkg/controller/handler.go:198]: ya, you are right, we should add it to TryMutateWorkflow

File: pkg/controller/handler.go [hunk 6]
```diff
@@ -180,40 +184,66 @@ func (p *Propeller) Handle(ctx context.Context, namespace, name string) error {
...
+	for streak = 0; streak < maxLength; streak++ {
+		t := p.metrics.RoundTime.Start(ctx)
+		mutatedWf, err := p.TryMutateWorkflow(ctx, w)
+		if err != nil {
+			// NOTE We are overriding the deepcopy here, as we are essentially ingnoring all mutations
+			// We only want to increase failed attempts and discard any other partial changes to the CRD.
+			mutatedWf = RecordSystemError(w, err)
+			p.metrics.SystemError.Inc(ctx)
+		} else if mutatedWf == nil {
+			return nil
```
Comment (EngHabu) [inline pkg/controller/handler.go:204]: Is this a legit case? that looks wrong... I know it existed... maybe at least add a log line here?

File: pkg/controller/handler.go [hunk 7]
```diff
@@ -180,40 +184,66 @@ func (p *Propeller) Handle(ctx context.Context, namespace, name string) error {
...
-			SetCompletedLabel(mutatedWf, time.Now())
-			ResetFinalizers(mutatedWf)
+		// TODO we will need to call updatestatus when it is supported. But to preserve metadata like (label/finalizer) we will need to use update
+
+		// update the GetExecutionStatus block of the FlyteWorkflow resource. UpdateStatus will not
+		// allow changes to the Spec of the resource, which is ideal for ensuring
+		// nothing other than resource status has been updated.
+		newWf, updateErr := p.wfStore.Update(ctx, mutatedWf, workflowstore.PriorityClassCritical)
+		if updateErr != nil {
+			return updateErr
```
Comment (EngHabu) [inline pkg/controller/handler.go:228]: Do you intentionally only want to measure the latency on success?

File: pkg/controller/workflowstore/inmemory.go (hunk 5 above)
Comment (EngHabu) [reply pkg/controller/workflowstore/inmemory.go:54]: Mind adding a comment to "NewInMemoryWorkflowStore" to expect this behavior?

**Review by EngHabu — COMMENTED** [review]
I left a couple of comments to write a bit more code docs...

File: pkg/controller/workflowstore/inmemory.go (hunk 5 above)
Comment (kumare3) [reply pkg/controller/workflowstore/inmemory.go:54]: sure

File: pkg/controller/handler.go (hunk 7 above)
Comment (kumare3) [reply pkg/controller/handler.go:228]: nope, adding a defer is not okayed by the linter as it thinks it is a resource leak

Comment (EngHabu) [reply pkg/controller/handler.go:230]: I agree with the linter...

**Review by EngHabu — APPROVED** [review]
